File: csharp/alphametics/Alphametics.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class Alphametics
{
    private static readonly int[] Digits = Enumerable.Range(0, 10).ToArray();

    public static IDictionary<char, int> Solve(string equation)
    {
        var (left, right) = Parse(equation);

        if (left.Any(c => c.Length > right.Length))
            throw new ArgumentException("Invalid equation");

        var leftFactors = Factors(left);
        var rightFactors = Factors(new[] {right});

        var letters = leftFactors.Keys.Union(rightFactors.Keys).ToHashSet();
        if (letters.Count > Digits.Length)
            throw new ArgumentException($"Equation may not have more than {Digits.Length} different letters");

        var nonZeroLetters = left.Append(right).Select(term => term[0]).ToHashSet();

        var solutions =
            from solution in GenerateSolutions(letters)
            where nonZeroLetters.All(c => solution[c] != 0L)
            where leftFactors.Sum(solution) == rightFactors.Sum(solution)
            select solution;

        return solutions.FirstOrDefault() ?? throw new ArgumentException("Invalid equation");
    }

    private static (string[] Left, string Right) Parse(string equation)
    {
        var leftAndRight = equation.Split(" == ");
        return (leftAndRight[0].Split(" + "), leftAndRight[1]);
    }

    private static IDictionary<char, long> Factors(IEnumerable<string> terms)
    {
        var factors = new Dictionary<char, long>();

        foreach (var term in terms)
        {
            var termRev = string.Concat(term.Reverse());
            for (int i = 0; i < termRev.Length; i++)
            {
                var letter = termRev[i];
                factors[letter] = factors.GetValueOrDefault(letter, 0L) + (long)Math.Pow(10, i);
            }
        }

        return factors;
    }

    private static long Sum(this IDictionary<char, long> factors, IDictionary<char, int> solution) =>
        factors.Sum(kv => kv.Value * solution[kv.Key]);

    private static IEnumerable<IDictionary<char, int>> GenerateSolutions(IReadOnlyCollection<char> letters) =>
        Digits
            .Permutations(letters.Count)
            .Select(values => letters.Zip(values).ToDictionary(kv => kv.First, kv => kv.Second));

    private static IEnumerable<T[]> Permutations<T>(this T[] input, int length)
    {
        if (length == 1)
            return input.Select(t => new[] {t});

        return input.Permutations(length - 1)
            .SelectMany(t => input.Where(e => !t.Contains(e)), (t1, t2) => t1.Concat(new[] {t2}).ToArray());
    }
}

File: fsharp/alphametics/Alphametics.fs
module Alphametics

type Equation =
    { Terms: string list }

    member this.Left = this.Terms.[..this.Terms.Length - 2]

    member this.Right = List.last this.Terms

    member this.LeftLetters = this.Left |> Seq.collect id |> Set.ofSeq

    member this.RightLetters = this.Right |> Set.ofSeq

    member this.Letters =
        Set.union this.LeftLetters this.RightLetters

type Solution = Map<char, int>

let private parse (puzzle: string) =
    let terms =
        puzzle.Split(" == ")
        |> Seq.collect (fun s -> s.Split(" + "))
        |> Seq.toList

    { Terms = terms }

let private factors terms =
    let letterFactors term =
        Seq.rev term
        |> Seq.mapi (fun i c -> c, pown 10 i |> int64)

    terms
    |> Seq.collect letterFactors
    |> Seq.groupBy fst
    |> Seq.map (fun (c, f) -> c, Seq.sumBy snd f)

let rec private combinations n l =
    match n, l with
    | _, [] -> []
    | 1, xs -> List.map (fun x -> [ x ]) xs
    | k, xs ->
        combinations (k - 1) xs
        |> List.collect (fun c -> List.except c xs |> List.map (fun x -> c @ [ x ]))

let private solutions (equation: Equation): Solution seq =
    combinations (Set.count equation.Letters) [ 0 .. 9 ]
    |> Seq.map (Seq.zip equation.Letters >> Map.ofSeq)

let private findSolution equation =
    let nonZeroLetters =
        equation.Terms |> Seq.map (Seq.head) |> Set.ofSeq

    let noTermsStartWithZero (solution: Solution) =
        nonZeroLetters
        |> Seq.forall (fun c -> solution.[c] <> 0)

    let leftFactors, rightFactors =
        factors equation.Left, factors [ equation.Right ]

    let leftSumEqualsRightSum (solution: Solution) =
        let sum =
            Seq.sumBy (fun (c, f) -> (int64 solution.[c]) * f)

        sum leftFactors = sum rightFactors

    solutions equation
    |> Seq.filter noTermsStartWithZero
    |> Seq.filter leftSumEqualsRightSum
    |> Seq.tryHead

let private validate (equation: Equation) =
    match equation with
    | eq when Set.count eq.Letters > 10 -> Error "Too many letters in equation"
    | eq when (Seq.map String.length >> Seq.max) eq.Left > String.length eq.Right ->
        Error "Left side of equation has terms longer than the answer"
    | eq -> Ok eq

let solve puzzle =
    let equation = parse puzzle

    match validate equation with
    | Error _ -> None
    | Ok eq -> findSolution eq
